File: MovieCatalogueV2/app/src/main/java/com/google/fdp/moviecataloguev2/MainActivity.java
package com.google.fdp.moviecataloguev2;

import android.content.Intent;
import android.os.Bundle;
import android.provider.Settings;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.SearchView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.MenuItemCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;

import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.fdp.moviecataloguev2.fragments.FavouriteFragment;
import com.google.fdp.moviecataloguev2.fragments.MovieFragment;
import com.google.fdp.moviecataloguev2.viewmodels.MovieViewModel;

import java.util.Objects;

public class MainActivity extends AppCompatActivity implements BottomNavigationView.OnNavigationItemSelectedListener {

    private BottomNavigationView navView;
    private MovieViewModel viewModel;
    private Fragment pageContent;
    public  String KEY_FRAGMENT = MovieFragment.MOVIE_KEY;
    public String TITLE_FRAGMENT = "title";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        navView = findViewById(R.id.navView);
        navView.setOnNavigationItemSelectedListener(this);

        viewModel = ViewModelProviders.of(this).get(MovieViewModel.class);

        if (savedInstanceState == null) {
            openFragment(MovieFragment.newInstance(MovieFragment.MOVIE_KEY, false));
        } else {
            KEY_FRAGMENT = savedInstanceState.getString(TITLE_FRAGMENT);
            openFragment(Objects.requireNonNull(getSupportFragmentManager().getFragment(savedInstanceState, KEY_FRAGMENT)));
        }

    }

    private void openFragment(Fragment fragment) {
        pageContent = fragment;
        getSupportFragmentManager().beginTransaction()
                .replace(R.id.flContainer, pageContent, fragment.getClass().getSimpleName()).commit();

    }

    private void searchMovie(String keyword) {
        viewModel.searchData(KEY_FRAGMENT, keyword);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.setting_menu, menu);

        SearchView searchView = (SearchView) MenuItemCompat.getActionView(menu.findItem(R.id.menu_search));
        searchView.setQueryHint(getResources().getString(R.string.search_hint));
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {

                return false;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                if (s.length() == 0) {
                    viewModel.updateItem(KEY_FRAGMENT);
                } else {
                    searchMovie(s);
                }
                return false;
            }
        });
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_language_setting:
                startActivity(new Intent(Settings.ACTION_LOCALE_SETTINGS));
                break;
            case R.id.menu_reminder_setting:
                startActivity(new Intent(this, SettingActivity.class));
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
        switch (menuItem.getItemId()) {
            case R.id.navigation_movie:
                KEY_FRAGMENT = MovieFragment.MOVIE_KEY;
                openFragment(MovieFragment.newInstance(MovieFragment.MOVIE_KEY, false));
                return true;
            case R.id.navigation_tv_show:
                KEY_FRAGMENT = MovieFragment.TV_SHOW_KEY;
                openFragment(MovieFragment.newInstance(MovieFragment.TV_SHOW_KEY, false));
                return true;
            case R.id.navigation_favorite:
                openFragment(new FavouriteFragment());
                return true;
            default:
                return false;
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        outState.putString(TITLE_FRAGMENT, KEY_FRAGMENT);
        getSupportFragmentManager().putFragment(outState, KEY_FRAGMENT, pageContent);
        super.onSaveInstanceState(outState);
    }
}

File: MovieCatalog/app/src/main/java/com/alfianh/moviecatalog/MainActivity.kt
package com.alfianh.moviecatalog

import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import android.view.Menu
import android.view.MenuItem
import com.google.android.material.bottomnavigation.BottomNavigationView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import com.alfianh.moviecatalog.fragments.FavoriteFragment
import com.alfianh.moviecatalog.fragments.MovieFragment
import com.alfianh.moviecatalog.fragments.MovieFragment.Companion.MOVIE_KEY
import com.alfianh.moviecatalog.fragments.MovieFragment.Companion.TV_SHOW_KEY
import com.alfianh.moviecatalog.viewmodels.MovieViewModel
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    navView.setOnNavigationItemSelectedListener(onNavigationItemSelectedListener)
    openFragment(MovieFragment.newInstance(MOVIE_KEY))
  }

  private fun openFragment(fragment: Fragment) {
    supportFragmentManager.beginTransaction()
      .replace(R.id.flContainer, fragment, fragment.javaClass.simpleName).commit()
  }

  private val onNavigationItemSelectedListener =
    BottomNavigationView.OnNavigationItemSelectedListener { item ->
      when (item.itemId) {
        R.id.navigation_movie -> {
          openFragment(MovieFragment.newInstance(MOVIE_KEY))
          return@OnNavigationItemSelectedListener true
        }
        R.id.navigation_tv_show -> {
          openFragment(MovieFragment.newInstance(TV_SHOW_KEY))
          return@OnNavigationItemSelectedListener true
        }
        R.id.navigation_favorite -> {
          openFragment(FavoriteFragment())
          return@OnNavigationItemSelectedListener true
        }
      }
      false
    }

  override fun onCreateOptionsMenu(menu: Menu?): Boolean {
    menuInflater.inflate(R.menu.setting_menu, menu)
    return super.onCreateOptionsMenu(menu)
  }

  override fun onOptionsItemSelected(item: MenuItem?): Boolean {
    when (item?.itemId) {
      R.id.menu_language_setting -> {
        startActivity(Intent(Settings.ACTION_LOCALE_SETTINGS))
      }
    }
    return super.onOptionsItemSelected(item)
  }

}
